kernel/kernel-api/src/main/java/io/delta/kernel/internal/checksum/ChecksumUtils.java

```diff
@@ -0,0 +1,162 @@
+    if (snapshot.getCurrentCrcInfo().isPresent()) {
+      return;
+    }
+
+    // TODO: Optimize using last available crc after https://github.com/delta-io/delta/pull/4112
+    LongAdder tableSizeByte = new LongAdder();
+    LongAdder fileCount = new LongAdder();
+    FileSizeHistogram fileSizeHistogram = FileSizeHistogram.createDefaultHistogram();
+    Map<String, DomainMetadata> domainMetadataMap = new HashMap<>();
+    ChecksumWriter checksumWriter = new ChecksumWriter(snapshot.getLogPath());
+
+    try (CreateCheckpointIterator checkpointIterator =
+        new CreateCheckpointIterator(
+            engine,
+            snapshot.getLogSegment(),
+            // Set minFileRetentionTimestampMillis to infinite future to skip all removed files
+            Instant.ofEpochMilli(Long.MAX_VALUE).toEpochMilli())) {
+
+      // Process all checkpoint batches
+      while (checkpointIterator.hasNext()) {
+        FilteredColumnarBatch filteredBatch = checkpointIterator.next();
+        ColumnarBatch batch = filteredBatch.getData();
+        Optional<ColumnVector> selectionVector = filteredBatch.getSelectionVector();
+        final int rowCount = batch.getSize();
+
+        ColumnVector removeVector = batch.getColumnVector(REMOVE_INDEX);
+        ColumnVector addVector = batch.getColumnVector(ADD_INDEX);
+        ColumnVector domainMetadataVector = batch.getColumnVector(DOMAIN_METADATA_INDEX);
+
+        // Process all selected rows in a single pass for optimal performance
+        for (int i = 0; i < rowCount; i++) {
+          // Check if this row is selected
+          boolean isSelected =
+              !selectionVector.isPresent()
+                  || (!selectionVector.get().isNullAt(i) && selectionVector.get().getBoolean(i));
+          if (!isSelected) continue;
+
+          // Step 1: Ensure there are no remove records
+          // We set minFileRetentionTimestampMillis to infinite future to skip all removed files,
+          // so there should be no remove actions.
+          checkState(
+              removeVector.isNullAt(i),
+              "unexpected remove row found when "
+                  + "setting minFileRetentionTimestampMillis to infinite future");
+
+          // Step 2: Process add file records - direct columnar access for better performance
+          if (!addVector.isNullAt(i)) {
+            // Direct access to size field in add file record for maximum performance
+            ColumnVector sizeVector = addVector.getChild(ADD_SIZE_INDEX);
+            long fileSize = sizeVector.getLong(i);
+
+            tableSizeByte.add(fileSize);
+            fileSizeHistogram.insert(fileSize);
+            fileCount.increment();
+          }
+
+          // Step 3: Process domain metadata records
+          if (!domainMetadataVector.isNullAt(i)) {
+            DomainMetadata domainMetadata =
+                DomainMetadata.fromColumnVector(domainMetadataVector, i);
+            if (domainMetadata != null
```

Comment:
We should only see 1 domain metadata selected for each domainName right?

Reply:
That's a good find https://github.com/delta-io/delta/blob/56cb730f0bd0639018d09c16546ea88c5211c2f3/kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/CreateCheckpointIterator.java#L387-L390 enforces that